Fix `encodeBase64` for empty password or user in HTTP Basic Authentication

Saibamen opened the pull request:
⚠️⚠️⚠️ Since we do not accept all types of pull requests and do not want to waste your time. Please be sure that you have read pull request rules:
https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md#can-i-create-a-pull-request-for-uptime-kuma

Tick the checkbox if you understand [x]:
- [x] I have read and understand the pull request rules.

# Description

Fix basic authorization for servers with empty password or empty user.

## Type of change

Please delete any options that are not relevant.

- Bug fix (non-breaking change which fixes an issue)

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I ran ESLint and other linters for modified files
- [x] I have performed a self-review of my own code and tested it
- [ ] I have commented my code, particularly in hard-to-understand areas (including JSDoc for methods)
- [x] My changes generates no new warnings
- [ ] My code needed automated testing. I have added them (this is optional task)

## Screenshots (if any)

### Proof of empty user auth with Postman

#### No auth - error 401

![image](https://github.com/louislam/uptime-kuma/assets/905878/4aec6241-b5ca-4efd-ad9e-000e51632b94)

#### Auth with empty user - 200 OK

![image](https://github.com/louislam/uptime-kuma/assets/905878/19f77a1e-2fbe-4fe2-97ec-2c3f46ac9bd2)

CommanderStorm commented on server/model/monitor.js:
```diff
@@ -233,7 +233,15 @@ class Monitor extends BeanModel {
      * @returns {string}
      */
     encodeBase64(user, pass) {
-        return Buffer.from(user + ":" + pass).toString("base64");
```
what is the type of `user`/`pass` in this context?
If it is an empty string, this should work as is, otherwise fixing the producer or adding   `|| ''` might be a better solution.

Saibamen replied:
> what is the type of user/pass in this context?
String

CommanderStorm commented:
I cannot reproduce (via https://webhook.site/) the current system not working.
(Also your reproduction steps in the description are using postman, not our system):

- `YWRtaW46` -> `admin:`
![image](https://github.com/louislam/uptime-kuma/assets/26258709/b31ae49a-2a98-43f2-910c-82c9cfdfa517)

- `YWRtaW46` -> `:admin`
  ![image](https://github.com/louislam/uptime-kuma/assets/26258709/65ff78f4-0593-4a54-8231-48248aededd5)

louislam commented:
Maybe off topic and maybe an edge case. If `user` contains a colon `:`, it properly will not be working.

Saibamen commented:
Commited a simplified version

Saibamen commented:
### Before changes:

![image](https://github.com/louislam/uptime-kuma/assets/905878/5f56d142-2fe5-4e42-b6f1-187748216842)

### After changes:

![image](https://github.com/louislam/uptime-kuma/assets/905878/1f93b630-5e0d-4816-9c18-7bb8c9cd1ac5)


Please reopen this bugfix.

HTTP Auth was set up for C# Server (Kestrel):

```csharp
public async Task InvokeAsync(HttpContext context)
{
    // Make sure we are hitting the swagger path, and not doing it locally as it just gets annoying :-)
    if (context.Request.Path.StartsWithSegments("/swagger") && !IsLocalRequest(context))
    {
        string authHeader = context.Request.Headers["Authorization"];
        if (authHeader != null && authHeader.StartsWith("Basic "))
        {
            // Get the encoded username and password
            var encodedUsernamePassword = authHeader.Split(' ', 2, StringSplitOptions.RemoveEmptyEntries)[1]?.Trim();

            // Decode from Base64 to string
            var decodedUsernamePassword = Encoding.UTF8.GetString(Convert.FromBase64String(encodedUsernamePassword));

            // Split username and password
            var username = decodedUsernamePassword.Split(':', 2)[0];
            var password = decodedUsernamePassword.Split(':', 2)[1];

            if (IsAuthorized(username, password))
            {
                await _next.Invoke(context);
                return;
            }
        }

        // Return authentication type (causes browser to show login dialog)
        context.Response.Headers["WWW-Authenticate"] = "Basic";

        // Return unauthorized
        context.Response.StatusCode = (int)HttpStatusCode.Unauthorized;
    }
    else
    {
        await _next.Invoke(context);
    }
}

public bool IsAuthorized(string username, string password)
{
    // Check that username and password are correct
    return username.Equals("", StringComparison.InvariantCultureIgnoreCase) && password.Equals("xxxxxxxxxxxxxxxxxx");
}
```

louislam commented:
Is it because the `user` is `undefined`, so it becomes `undefined:yourPassword`?

Saibamen commented:
`null`:
![image](https://github.com/louislam/uptime-kuma/assets/905878/38a72699-176f-4f31-9c4a-5c23712beb56)

CommanderStorm commented:
The reason, why I could not reproduce this is that I tested with set pw/username first.

Initially these values are `null`
![image](https://github.com/louislam/uptime-kuma/assets/26258709/1364a3ba-7552-476d-ad47-772a99668613)

but after edting (adding+removing a char) these values become `""`
![image](https://github.com/louislam/uptime-kuma/assets/26258709/a91de333-0498-44dd-a799-0c54d7333553)

CommanderStorm commented:
Please include a type hint here
https://github.com/louislam/uptime-kuma/pull/4326/files#diff-0c4874f149ad01732ad730f36c95c99e37f5c3f13e9b52f628e2e284daa72d27R232

like 

```js
/**
 * Encode user and password to Base64 encoding
 * for HTTP "basic" auth, as per RFC-7617
 * @param {string|null} user - The username (nullable if not changed by a user)
 * @param {string|null} pass - The password (nullable if not changed by a user)
 * @returns {string}
 */

CommanderStorm commented:
@Saibamen 

> [!NOTE]
> This function in your code is not resistent to timing attacks
>
> ```
> public bool IsAuthorized(string username, string password)
> {
>     // Check that username and password are correct
>     return username.Equals("", StringComparison.InvariantCultureIgnoreCase) && password.Equals("xxxxxxxxxxxxxxxxxx");
> }
> ```
> 
> See CWE-385 for further details: https://cwe.mitre.org/data/definitions/385.html
